README.md
# Catalyst::Plugin::BigSitemap

## Version 1.0.1

## Change History 

### 1.0.1

* Documentation updates
* MAJOR CHANGE: SitemapBuilder public interface no longer includes a urls array.  With a large quantity of urls (hundreds of millions)
  using this accessor would likely cause a system crash by placing all the URI objects in an arrayref.
* MAJOR CHANGE: Removed type constraint for URLs to be of type URI::Http.  Any valid URI can now be used.

### 1.0

* Converted SitemapBuilder to cache URL data to a SQLite database instead of storing in RAM.

### 0.9

* Unit tests written 
* Minor bugfix in SitemapBuilder->\_urls\_slice method.  

### 0.02

* Added **sitemap** and **sitemap_as_xml** attributes to the plugin module to make for complete interface compatability with **Catalyst::Plugin::Sitemap 1.0.0**

### 0.01

* Initial version

## Description

A [Catalyst Framework](http://catalystframework.org) Plugin that allows for automatic generation, and caching to disk
of [Sitemap](http://sitemaps.org/protocol.html) and [Sitemap Index](http://sitemaps.org/protocol.html#index) files to support
beyond the 50,000 URL max of a single sitemap, to a maximum of 2.5 billion urls.

```
cpan install Catalyst::Plugin::BigSitemap
```

## Synopsis

### MyCatalystApp.pm
```
use Catalyst qw/BigSitemap/;
```

### MyCatalystApp.conf (Config::General flavored)
```
<Plugin::BigSitemap>
    cache_dir /var/www/myapp/root/sitemaps
    url_base http://mywebsite/
    sitemap_name_format sitemap%d.xml.gz
    sitemap_index_name sitemap_index.xml
</Plugin::BigSitemap>
```

### MyApacheConf.conf

Assuming you have [mod_alias](http://httpd.apache.org/docs/2.2/mod/mod_alias.html#aliasmatch) installed
and you want to store your sitemap files on the disk and serve them straight through apache (Seriously.. 
if your sitemaps are large enough to warrant using this module, then you definitely don't want to be 
building and serving from Catalyst for each request..)

```
<VirtualHost *.80>
    ... your standard configuration ... 
    Alias /sitemap_index.xml /var/www/mysite/root/sitemaps/sitemap_index.xml
    AliasMatch ^/sitemap(\d+).xml.gz$ /var/www/mysite/root/sitemaps/sitemap$1.xml.gz
</VirtualHost>
```

### MyController.pm
```
#
# Actions you want included in your sitemap.  In this example, there's a total of 10 urls that will be written
#

sub single_url_action :Local :Args(0) :Sitemap() { ... }
sub single_url_with_attrs : Local :Args(0) :Sitemap( loc => 'http://www.mysite/here', changefreq => 'daily', priority => '0.5' ) { ... }

sub multiple_url_action :Local :Args(1) :Sitemap('*') { ... }    
sub multiple_url_action_sitemap {
    my ( $self, $c, $sitemap ) = @_;
    
    # just add 8 more arbitrary urls
    my $a = $c->controller('MyController')->action_for('multiple_url_action');
    for (my $i = 0; $i < 8; $i++) {
        my $uri = $c->uri_for($a, [ $i, ]);
        $sitemap->add( $uri );
    }
    
}

#
# Action to rebuild your sitemaps and writes them to your harddisk 
# !!! You want to protect this !!!
# Best thing to do would be manually instantiate an instance of your
# application from the cron job, mark this method private and call it.  
# You could also go crazy and use WWW::Mechanize .. or hell.. leave it
# public and call it from your browser.. your call.  I wouldn't do that, 
# though ;) 
# Your old sitemap files will automatically be overwritten each time this
# is called.
#

sub rebuild_cache :Private {
    my ( $self, $c ) = @_;
    $c->write_sitemap_cache();
}

#
# Serving the sitemap files is best to do directly through apache.. 
# New version of catalyst have depreciated regex actions, which
# makes doing sitemap files a little more difficult (though you
# can still manually include support for regex actions)
# 
# Also, if you only have a single sitemap, and want to use this like 
# Catalyst::Plugin::Sitemap, see sub single_sitemap below. 
#

sub sitemap_index :Private {
    my ( $self, $c ) = @_;
    
    my $smi_xml = $c->sitemap_builder->sitemap_index->as_xml;
    $c->response->body( $smi_xml );
}

sub single_sitemap :Private {
    my ( $self, $c ) = @_;
    
    my $sm_xml = $c->sitemap_builder->sitemap(0)->as_xml;
    $c->response->body( $sm_xml );
}
```

## Note

This is designed to _almost_ be a drop-in replacement for the existing [Catalyst::Plugin::Sitemap](https://metacpan.org/module/Catalyst::Plugin::Sitemap), 
and the URL attributes work the exact same way.  Your controller actions are attributed in the EXACT same way.  
The only difference is when you want to serve your sitemap file.

## TODO List

* Allow for lastmodified on SitemapIndex files.  



CHANGES
1.0.1
	- Bugfix (major): Fixed a problem caused when adding a very large number of URLs to a sitemap.  
		All the URIs for SitemapBuilder were stored in RAM.  Now a temp sqlite database is created,
		and URI data is stored there.  The database is deleted after your script exits.	

0.9.1
	- Bugfix (major):  Fixed error that caused type validation error when 'base_url' was overridden in site configuration.
		The value read out of the configuration file was not being properly converted to a URI object, instead it was
		passed to SitemapBuilder as a string, which failed validation.
0.9
    - Feature: Added unit tests
    - Bugfix (minor): SitemapBuilder->_urls_slice wasn't returning expected slices of URLs array.  Resolved.
    
0.02
    - Feature: Added two methods to BigSitemap.pm for complete interface compatibility with Sitemap.pm
        - sitemap
        - sitemap_as_xml
        
0.01      
    - Initial Release

t/lib/TestApp/Controller/Root.pm
package TestApp::Controller::Root;

use strict;
use warnings;

use parent 'Catalyst::Controller';

sub sitemap : Path('/sitemap') {
    my ( $self, $c ) = @_;
    $c->res->body( $c->sitemap_builder->sitemap(0)->as_xml );
}

sub dynamic :Path('/dynamic') {
    my ( $self, $c ) = @_;

    $c->sitemap->add( "http://localhost/sumfin" ); #''.$c->uri_for( '/sumfin' ) );

    $c->res->body( 'dynamic it is' );
}

sub alone :Local :Sitemap { }

sub with_priority :Local :Sitemap(0.75) { }

sub with_function :Local :Sitemap(*) { }

sub with_function_sitemap {
    $_[2]->add( 'http://localhost/root/with_function' );
}

sub with_args :Local 
    :Sitemap( lastmod => 2010-09-27, changefreq => daily ) 
    {}


1;








lib/Catalyst/Plugin/BigSitemap/SitemapBuilder.pm
package Catalyst::Plugin::BigSitemap::SitemapBuilder;
use Modern::Perl '2010';
use WWW::Sitemap::XML;
use WWW::Sitemap::XML::URL;
use WWW::SitemapIndex::XML;
use Carp;
use Try::Tiny;
use Data::Dumper;
use DBI;
use DBD::SQLite;
use File::Temp qw/tempdir/;
use Path::Class;
use Moose;

BEGIN { our $VERSION = '1.0'; }

=head1 NAME 

Catalyst::Plugin::BigSitemap::SitemapBuilder - Helper object for the BigSitemap plugin

=head1 VERSION

1.0

=head1 DESCRIPTION

This object's role is to accept a collection of L<WWW::Sitemap::XML::URL> objects via the L<add>
method.  

=head1 CONSTRUCTOR

There are two required parameters that must be passed to the constructor, L<sitemap_base_uri> and
L<sitemap_name_format>.  


=head1 ATTRIBUTES

=shift 4

=item urls - I<ArrayRef> of L<WWW::Sitemap::XML::URL>

A collection of every URL in your application that will be included in the sitemap.

=item sitemap_base_uri - L<URI::http>

The base URI that should be used when resolving the action urls in your application.  
You should really specify this manually, in the event that one day you want to start
run this module from a cron job.

=item sitemap_name_format - I<Str>

A sprintf style format for the names of your sitemap files.  Note:  The names of the sitemap files
will start by inserting the number 1 and incrementing for each sitemap file written.  It's important
to note that in code, calls to the sitemap method use a 0-based-index but your sitemap filenames are
1-based.  This is just that way so the names of the individual sitemaps match to the examples given
on the L<http://www.sitemaps.org> website.

=item failed_count - I<Int>

A running count of all the URLs that failed validation in the L<WWW::Sitemap::XML::URL> module and could not 
be added to the collection.. This should always report zero unless you've screwed something up in your
C<sub my_action_sitemap> controller methods.

=back

=head1 TODOs

lastmod is not supported on the sitemap index.

=cut

has 'sitemap_base_uri'      => ( is => 'ro', isa => 'URI' );
has 'sitemap_name_format'   => ( is => 'ro', isa => 'Str' );
has 'failed_count'          => ( is => 'rw', isa => 'Int', default => 0 );
has 'dbh'                   => ( is => 'ro', builder => '_build_dbh' );
has '_insert_cmd'           => ( is => 'ro', builder => '_build_insert_cmd', lazy => 1 );

=head1 METHODS

=over 4

=item add( $myUrlString )
=item add( $myUriObject )
=item add( loc => ? [, changefreq => ?] [, priority => ?] [, lastmod => ?] )

This method comes in three flavors.  The first, take a single string parameter that should be the stringified version of the
URL you want to add to the sitemap. The second, takes a URI::http object.  The last flavor takes a hashref containing all your
input parameters. 

=item urls_count() = Int

.. how many urls total have been added to the builder.

=item sitemap_count() - Int

.. how many total sitemap files can be built with this data.

=item sitemap_index() - L<WWW::SitemapIndex::XML>

Generates and returns a new sitemapindex object based on the urls currently in this object's
urls collection, the sitemap_base_uri and the sitemap_name_format setting.  

=item sitemap($index) - L<WWW::Sitemap::XML>

Generates and returns a new sitemap object based at your requested index.

B<Note:> $index is a 0-based index of the sitemap you want to retrieve. 

=back

=cut

sub add {
    my $self = shift;
    my @params = @_;
    
    # create our url object.. for compatability with Catalyst::Plugin::Sitemap
    # we allow a single string parameter to be passed in.
    #
    # NOTE: The WWW::Sitemap::XML::URL object is immediately discarded afterwards.  This is 
    # a quick way to make sure it validates properly before we put it in the database.
    my $u;
    try {    	
        if (@params == 0) {
            croak "method add() requires at least one argument.";
        }
        elsif (@params == 1){  
        	# if only a single parameter is provided, we assume it's location
            my $loc = URI->new($params[0]);
            my $host = URI->new( $self->sitemap_base_uri );
            $loc->host( $host->host );
 
            $u = WWW::Sitemap::XML::URL->new(loc => $loc);
        }
        elsif (@params % 2 == 0) {       
            # otherwise, we need an even number of args
            my %ph = @params;      
            my $loc = URI->new($ph{loc});           
            my $host = URI->new( $self->sitemap_base_uri );
            $loc->host( $host->host );
            $ph{loc} = $loc;
            $u = WWW::Sitemap::XML::URL->new(%ph);
        }        
        else {                        
            croak "method add() requires either a single argument, or an even number of arguments.";  
        }
        
        
        $self->_insert_cmd->execute($u->loc, $u->lastmod, $u->changefreq, $u->priority);        
    }
    catch {           
        $self->failed_count($self->failed_count + 1);
    };
    
}

sub urls_count {
    my $self = shift;
    
    return $self->dbh->selectrow_hashref('SELECT COUNT(*) AS c FROM uri')->{c};
}

sub sitemap_count {
    my $self = shift;
    
    my $whole_pages     = int ( $self->urls_count / 50_000 );
    my $partial_pages   = $self->urls_count % 50_000 ? 1 : 0; 
    
    return $whole_pages + $partial_pages;    
}

sub sitemap_index {
    my $self = shift;
    
    my $smi = WWW::SitemapIndex::XML->new();
    
    for (my $index = 0; $index < $self->sitemap_count; $index++) {   
        # TODO: support lastupdate
        $smi->add( loc => $self->sitemap_base_uri->as_string . sprintf($self->sitemap_name_format, ($index + 1)) );
    }
    
    return $smi;    
}

sub sitemap {
    my ( $self, $index ) = @_;    
    
    my @sitemap_urls = $self->_urls_slice( $index );
    
    my $sm = WWW::Sitemap::XML->new();
    
    foreach my $url (@sitemap_urls) {
        try{            
            $sm->add($url);    
        }
        catch{
            warn "Problem adding url to sitemap: " . Dumper $url;    
        };
    }
    
    return $sm;    
}

=head1 INTERNAL USE METHODS

Methods you're not meant to use directly, so don't!  They're here for documentation
purposes only.

=over 4

=item _urls_slice($index)

Returns an array slice of URLs for the sitemap at the provided index.  
Sitemaps can consist of up to 50,000 URLS, when creating the slice, 
we use the assumption that we'll try to get up to 50,000 per each 
sitemap.

=back

=cut

sub _urls_slice {
    my ( $self, $index ) = @_;
    
    my $start_index = $index * 50_000;
    
    my $rows = $self->dbh->selectall_hashref("SELECT * FROM uri LIMIT 50000 OFFSET $start_index", 'id');
    my @urls = ();
    foreach my $id (keys %$rows) {
        my $d = $rows->{$id};
        
        my %constructor_args = (loc => $d->{loc});
        $constructor_args{lastmod} = $d->{lastmod} if $d->{lastmod};
        $constructor_args{changefreq} = $d->{changefreq} if $d->{changefreq};
        $constructor_args{priority} = $d->{priority} if $d->{priority};
        
        push @urls, WWW::Sitemap::XML::URL->new(%constructor_args);
    }
    
    return @urls;
}

=head1 BUILDER METHODS

=over 4

=item dbh_builder()

=back

=cut

sub _build_dbh {
    my $self = shift;
    
    # create + connect to temporary database
    my $db_path = Path::Class::File->new(tempdir(CLEANUP => 1), 'catalyst-plugin-bigsitemap.db')->stringify();    
    my $dbh = DBI->connect(
                'dbi:SQLite:dbname=' . $db_path,
                '',''
              ) || die 'Could not create database';
              
    # make our single table to use
    $dbh->do('CREATE TABLE uri (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                loc TEXT NOT NULL,
                lastmod TEXT NULL,
                changefreq TEXT NULL,
                priority TEXT NULL
              )');
    
    return $dbh;    
}

sub _build_insert_cmd {
    my $self = shift;    
    return $self->dbh->prepare('INSERT INTO uri (loc, lastmod, changefreq, priority) VALUES (?,?,?,?)');
}



=head1 SEE ALSO

=head1 AUTHOR

Derek J. Curtis C<djcurtis at summersetsoftware dot com>

=head1 COPYRIGHT

Derek J. Curtis 2013

=head1 LICENSE

This library is free software. You can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

1;


lib/Catalyst/Plugin/BigSitemap.pm
package Catalyst::Plugin::BigSitemap;
use Modern::Perl '2010';
use Catalyst::Plugin::BigSitemap::SitemapBuilder;
use WWW::SitemapIndex::XML;
use WWW::Sitemap::XML;
use Path::Class;
use Carp;
use Moose;
use URI;


BEGIN { our $VERSION = '1.0'; }

=encoding utf8

=head1 NAME 

Catalyst::Plugin::BigSitemap - Auto-generated Sitemaps for up to 2.5 billion URLs.

=head1 VERSION

1.0

=head1 DESCRIPTION

A compete drop-in replacement for L<Catalyst::Plugin::Sitemap> but gives you the 
following features as well.

Additional Functionality:

1. Automatic generation of a sitemap index with multiple sitemap files.

2. Writing sitemap files to disk as XML (can be gzipped as well).

3. Configurable naming of sitemap and sitemap index files.  

4. 2,500,000,000 URL maximum

Even if you don't have a site with more than 50,000 URLS, this module can
write (and rewrite) your sitemap files to disk (something the original L<Catalyst::Plugin::Sitemap> 
module lacks), so you may wish to consider this module for no other reason than that.

=head1 SYNOPSIS

    #
    # Actions you want included in your sitemap.  In this example, there's a total of 10 urls that will be written
    #

    # loc will be automatically resolved for this action
    sub single_url_action :Local :Args(0) :Sitemap() { ... }
    
    # see WWW::Sitemap::XML::URL for all the possible attributes 
    sub single_url_with_attrs : Local :Args(0) :Sitemap( changefreq => 'daily', priority => '0.5' ) { ... }
    
    # If you have multiple urls that map to an action, you need to specify the string '*' as it's only attribute
    # then you must specify a method with the same name plus '_sitemap' as your action method.  Within that action,
    # you'll need to call the ->add method on your sitemap.  
    sub multiple_url_action :Local :Args(1) :Sitemap('*') { ... }    
    sub multiple_url_action_sitemap {
        my ( $self, $c, $sitemap ) = @_;
        
        my $a = $c->controller('MyController')->action_for('multiple_url_action');
        for (my $i = 0; $i < 8; $i++) {
            my $uri = $c->uri_for($a, [ $i, ]);
            $sitemap->add( $uri );
        }
        
    }

    #
    # Action to rebuild your sitemap -- you want to protect this!
    # Best thing to do would be manually instantiate an instance of your
    # application from the cron job, mark this method private and call it.  
    # You could also go crazy and use WWW::Mechanize .. or hell.. leave it
    # public and call it from your browser.. your call.  I wouldn't do that, 
    # though ;) 
    # Your old sitemap files will automatically be overwritten.  
    #
    
    sub rebuild_cache :Private {
        my ( $self, $c ) = @_;
        $c->write_sitemap_cache();
    }
    
    #
    # Serving the sitemap files is best to do directly through apache.. 
    # New version of catalyst have depreciated regex actions, which
    # makes doing sitemap files a little more difficult (though you
    # can still manually include support for regex actions)
    # 
    # Also, if you only have a single sitemap, and want to use this like 
    # Catalyst::Plugin::Sitemap, see sub single_sitemap below. 
    #
    
    sub sitemap_index :Private {
        my ( $self, $c ) = @_;
        
        my $smi_xml = $c->sitemap_builder->sitemap_index->as_xml;
        $c->response->body( $smi_xml );
    }
    
    sub single_sitemap :Private {
        my ( $self, $c ) = @_;
        
        my $sm_xml = $c->sitemap_builder->sitemap(0)->as_xml;
        $c->response->body( $sm_xml );
    }


=head1 CONFIGURATION

There are a few configuration settings that must be set for this application to function properly.

=over 4

=item cache_dir - B<required>

The absolute filesystem path to where your sitemap file will be written when calling $c->write_sitemap_cache()

=item url_base - I<optional: defaults to whichever base url the request is made to>

This is the base url that will be used when building the urls for your application.

B<Note:> This is important especially if your rebuild is being launched by a cronjob that's
making a request to localhost.  In that case, if you fail the specify this setting, all your
urls will be resolved to http://localhost/my-action-here/ ... That probably isn't what you want.

B<Note:> The trailing slash is important!

=item sitemap_name_format - I<optional: defaults to sitemap%d.xml.gz>

A L<sprintf> format string.  Your sitemaps will be named beginning with 1 up through the total 
number of sitemaps that are necessary to build your data.  By default, this will end up being
something like 

B<Note:> The file extension should either be C<.xml> or C<.xml.gz>.  The proper type of file will be 
built depending on which extension you specify.

=item sitema_index_name - I<optional: defaults to sitemap_index.xml>

B<Note:> Just like with sitename_name_format, .xml or .xml.gz should be specified as the file 
extension.

=back

=head2 L<Config::General> Example

    <Plugin::BigSitemap>
        cache_dir /var/www/myapp/root/sitemaps
        url_base http://mywebsite/
        sitemap_name_format sitemap%d.xml.gz
        sitemap_index_name sitemap_index.xml
    </Plugin::BigSitemap>

=head1 ATTRIBUTES

=over 4

=item sitemap_builder

A lazy-loaded L<Catalyst::Plugin::BigSitemap::SitemapBuilder> object.  If you want access to the individual
L<WWW::Sitemap::XML> or the L<WWW::SitemapIndex::XML> file, you'll do that through this object.

=item sitemap 

B<Provided for compatability with Catalyst::Plugin::Sitemap.>  

Returns a L<WWW::Sitemap::XML> file containing
up to the first 50,000 URLs resolved from your application.

=item sitemap_as_xml

B<Provided for compatability with Catalyst::Plugin::Sitemap>.

Returns a L<XML::LibXML::Document> representation of the sitemap
generated by this module's L<sitemap> attribute.

=back

=cut

has 'sitemap_builder'   => ( is => 'rw', builder => '_get_sitemap_builder', lazy => 1, );
has 'sitemap'           => ( is => 'ro', builder => '_get_sitemap', lazy => 1, );
has 'sitemap_as_xml'    => ( is => 'ro', builder => '_get_sitemap_as_xml', lazy => 1, );

=head1 METHODS

=over 4

=item write_sitemap_cache()

Writes your sitemap_index and sitemap files to whichever cache_dir you've specified in your configuration.

On success, returns an array with the absolute path to the sitemap index (element 0), and all sitemap files.

=back

=cut

sub write_sitemap_cache {
    my $self = shift;

    my $cache_dir   = dir( $self->config->{'Plugin::BigSitemap'}->{cache_dir} || croak 'cache_dir setting MUST be specified in your app\'s config' );
    
    my $sitemap_index_filename = $self->config->{'Plugin::BigSitemap'}->{sitemap_index_name} || 'sitemap_index.xml';
    my $sitemap_index_full_name = $cache_dir->file($sitemap_index_filename)->stringify;    
    
    $self->sitemap_builder->sitemap_index->write( $sitemap_index_full_name );         
    
    my @sitemaps_written;
    for (my $i = 0; $i < $self->sitemap_builder->sitemap_count; $i++) {  
              
        my $filename    = sprintf($self->sitemap_builder->sitemap_name_format, ($i + 1));        
        my $full_name   = $cache_dir->file($filename)->stringify;        
    
        $self->sitemap_builder->sitemap($i)->write( $full_name );
        
        push @sitemaps_written, $full_name;        
    }
    
    return ( $sitemap_index_full_name, @sitemaps_written );
}

=head1 INTERNAL USE METHODS

Methods you shouldn't be calling directly.. They're listed here for documentation purposes.

=over 4

=item _get_sitemap_builder()

Returns a sitemap builder object that's fully populated with all the sitemap urls registered.
This can take quite some time depending on the number of urls you're registering with the sitemap
and how they're being generated.  

You shouldn't ever need to call this directly -- it's set as the builder method for the L<sitemap_builder> attribute.

B<Note>:  This can take an incredibly long time especially if you have a lot of URLs!  Use with care!

=item _get_sitemap()

Builder method for sitemap attribute. 

=item _get_sitemap_as_xml()

Builder method for sitemap_as_xml attribute.

=back

=cut

sub _get_sitemap_builder {
    my $self = shift;
    
    # setup our builder
    my $sb = Catalyst::Plugin::BigSitemap::SitemapBuilder->new(    
        sitemap_base_uri    => URI->new( $self->config->{'Plugin::BigSitemap'}->{url_base} )  || $self->req->base,
        sitemap_name_format => $self->config->{'Plugin::BigSitemap'}->{sitemap_name_format} || 'sitemap%d.xml.gz',
    );
   
    $sb->dbh->begin_work;
    say "Starting build process at " . DateTime->now();
    

    # Ugly ugly .. but all we're doing here is looping over every action of every controller in our application.
    CONTROLLER:
    foreach my $controller ( map { $self->controller($_) } $self->controllers ) {          

        ACTION: 
        foreach my $action ( map { $controller->action_for( $_->name ) } $controller->get_action_methods ) {

            say 'Getting urls for action: ' . $action;

            # Make sure there's at least one sitemap action .. 
            # Throw an exception if there's more than one sitemap attribute
            my $attr = $action->attributes->{Sitemap} or next ACTION;
            croak "more than one attribute 'Sitemap' for sub " if @$attr > 1;

            my @attr = split /\s*(?:,|=>)\s*/, $attr->[0];

            my %uri_params;

            if ( @attr == 1 ) {                
                if ( $attr[0] eq '*' ) {
                    my $sitemap_method = $action->name . "_sitemap";

                    if ( $controller->can($sitemap_method) ) {
                        $controller->$sitemap_method( $self, $sb );
                        next ACTION;
                    }
                }

                if ( $attr[0] + 0 > 0 ) {
                    # it's a number 
                    $uri_params{priority} = $attr[0];
                }
            }
            elsif ( @attr > 0 ) {
                %uri_params = @attr;
            }

            $uri_params{loc} = $self->uri_for_action( $action->private_path );
            $sb->add(%uri_params);

            # commit to the database after every action's urls are resolved
            $sb->dbh->commit;
            $sb->dbh->begin_work;
            print '...total urls: ';
            
        } # foreach $action             
    } # foreach $controller
    
    $sb->dbh->commit;
    say "Sitemap builder complete at " . DateTime->now();
    say "URLs:     " . $sb->urls_count();
    say "Errors:   " . $sb->failed_count();
    say "Sitemaps: " . $sb->sitemap_count();
    return $sb;
}

sub _get_sitemap { $_[0]->sitemap_builder->sitemap(0) }

sub _get_sitemap_as_xml { $_[0]->sitemap->as_xml }

=head1 ACKNOWLEDGEMENTS

This module is based on the great work of two other CPAN authors:

L<< Yanick Champoux|http://search.cpan.org/~yanick/ >> - Author of the original L<Catalyst::Plugin::Sitemap>
module.  Some of his ideas (and some of his module code) was used to make this module.

L<< Alex J. G. Burzyński|http://search.cpan.org/~ajgb/ >> - Author of L<WWW::Sitemap::XML>, which is the underlying
module used by BigSitemap to do it's work.

=head1 SEE ALSO

L<Catalyst::Plugin::Sitemap>, L<WWW::Sitemap::XML>, L<< Sitemaps.org|http://www.sitemaps.org/ >>, L<< Catalyst Framework|http://catalystframework.org/ >>

=head1 AUTHOR

Derek J. Curtis C<< <djcurtis at summersetsoftware dot com>> >>

Summerset Software, LLC

L<http://www.summersetsoftware.com>

=head1 LICENSE AND COPYRIGHT

Copyright 2013 Derek J. Curtis.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1;
